mqtt/mqtt-edgehub/src/command.rs

```diff
@@ -0,0 +1,266 @@
+use std::{collections::HashSet, iter::FromIterator, time::Duration};
+
+use futures_util::future::BoxFuture;
+use lazy_static::lazy_static;
+use regex::Regex;
+use tokio::{net::TcpStream, stream::StreamExt};
+use tracing::{error, info, warn};
+
+use mqtt3::{
+    proto, Client, Event, IoSource, ShutdownError, SubscriptionUpdateEvent, UpdateSubscriptionError,
+};
+use mqtt_broker::{BrokerHandle, Error, Message, SystemEvent};
+
+const TOPIC_FILTER: &str = "$edgehub/+/disconnect";
+const CLIENT_EXTRACTION_REGEX: &str = r"\$edgehub/(.*)/disconnect";
+
+#[derive(Debug)]
+pub struct ShutdownHandle(mqtt3::ShutdownHandle);
+
+impl ShutdownHandle {
+    pub async fn shutdown(&mut self) -> Result<(), Error> {
+        self.0.shutdown().await.map_err(Error::ShutdownClient)?;
+        Ok(())
+    }
+}
+
+pub struct BrokerConnection {
+    address: String,
+}
+
+impl IoSource for BrokerConnection {
+    type Io = TcpStream;
+    type Error = std::io::Error;
+    type Future = BoxFuture<'static, Result<(TcpStream, Option<String>), std::io::Error>>;
+
+    fn connect(&mut self) -> Self::Future {
+        let address = self.address.clone();
+        Box::pin(async move {
+            let io = TcpStream::connect(address).await;
+            io.map(|io| (io, None))
+        })
+    }
+}
+
+pub struct CommandHandler {
+    broker_handle: BrokerHandle,
+    client: Client<BrokerConnection>,
+}
+
+impl CommandHandler {
+    pub fn new(broker_handle: BrokerHandle, address: String, device_id: &str) -> Self {
+        let client_id = format!("{}/$edgeHub/$broker", device_id);
+
+        let client = Client::new(
+            Some(client_id),
+            None,
+            None,
+            BrokerConnection { address },
+            Duration::from_secs(1),
+            Duration::from_secs(60),
+        );
+
+        CommandHandler {
+            broker_handle,
+            client,
+        }
+    }
+
+    pub fn shutdown_handle(&self) -> Result<ShutdownHandle, ShutdownError> {
+        self.client
+            .shutdown_handle()
+            .map_or(Err(ShutdownError::ClientDoesNotExist), |shutdown_handle| {
+                Ok(ShutdownHandle(shutdown_handle))
+            })
+    }
+
+    pub async fn run(mut self) -> Result<(), CommandHandlerError> {
+        let subscribe_topics = &[TOPIC_FILTER.to_string()];
+
+        if !self.subscribe(subscribe_topics).await? {
+            return Err(CommandHandlerError::MissingSubacks(
+                subscribe_topics.concat(),
+            ));
+        }
+
+        while let Some(event) = self
+            .client
+            .try_next()
```

Comment:
nit: just curious why try_next? in what circumstances poll'ing the client can panic?

Reply:
I'm not sure why it could panic, but thought this was safer